library/core/src/num/mod.rs

```diff
@@ -846,37 +870,70 @@ fn from_str_radix<T: FromStrRadixHelper>(src: &str, radix: u32) -> Result<T, Par
     };
 
     let mut result = T::from_u32(0);
-    if is_positive {
-        // The number is positive
-        for &c in digits {
-            let x = match (c as char).to_digit(radix) {
-                Some(x) => x,
-                None => return Err(PIE { kind: InvalidDigit }),
-            };
-            result = match result.checked_mul(radix) {
-                Some(result) => result,
-                None => return Err(PIE { kind: PosOverflow }),
-            };
-            result = match result.checked_add(x) {
-                Some(result) => result,
-                None => return Err(PIE { kind: PosOverflow }),
-            };
+    if radix <= 16
+        && src.len()
+            <= mem::size_of::<T>() * 2 - if is_signed_ty { is_positive as usize } else { 0 }
```

Comment:
```suggestion
            <= mem::size_of::<T>() * 2 - (is_signed_ty && is_positive) as usize
```

they should be the same but this looks shorter 🙃